### PR TITLE
Spring Boot Portfolio Project — Part 4

### PR DESCRIPTION
**Pull request by mgucma**

### CODE REVIEW

File: src/main/java/com/marek/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+public class AuthorSpecificationProvider implements SpecificationProvider<Book> {
+    private static final String FIELD_NAME = "author";
+
+    @Override
+    public String getKey() {
+        return FIELD_NAME;
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return (root, query, criteriaBuilder) -> root.get("author")
```
Comment (beatarjn) [inline src/main/java/com/marek/onlinebookstore/repository/book/spec/AuthorSpecificationProvider.java]: You've created a constant for 'author', so let's use it.

File: src/main/java/com/marek/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+public class IsbnSpecificationProvider implements SpecificationProvider<Book> {
+    private static final String FIELD_NAME = "isbn";
+
+    @Override
+    public String getKey() {
+        return FIELD_NAME;
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return ((root, query, criteriaBuilder) -> root.get("isbn")
```
Comment (beatarjn) [inline src/main/java/com/marek/onlinebookstore/repository/book/spec/IsbnSpecificationProvider.java]: You've created a constant for 'isbn', so let's use it.

File: src/main/java/com/marek/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+public class TitleSpecificationProvider implements SpecificationProvider<Book> {
+    private static final String FIELD_NAME = "title";
+
+    @Override
+    public String getKey() {
+        return FIELD_NAME;
+    }
+
+    public Specification<Book> getSpecification(String[] params) {
+        return ((root, query, criteriaBuilder) -> root.get("title")
```
Comment (beatarjn) [inline src/main/java/com/marek/onlinebookstore/repository/book/spec/TitleSpecificationProvider.java]: You've created a constant for 'title', so let's use it.

File: src/main/java/com/marek/onlinebookstore/service/impl/BookServiceImpl.java [hunk 4]
```diff
@@ -52,4 +56,11 @@ public BookDto update(Long id, CreateBookRequestDto book) {
         modelBook.setId(id);
         return bookMapping.toBookDto(bookRepository.save(modelBook));
     }
+
+    @Override
+    public List<BookDto> searchBooks(BookSearchParametersDto searchParameters) {
+        Specification<Book> bookSpecification = bookSpecificationBuilder.build(searchParameters);
+        return bookRepository.findAll(bookSpecification)
+                .stream().map(bookMapping::toBookDto).toList();
```
Comment (beatarjn) [inline src/main/java/com/marek/onlinebookstore/service/impl/BookServiceImpl.java]: Format to the stream look:
    .stream()
    .map()
    ...

File: src/main/java/com/marek/onlinebookstore/repository/book/BookSpecificationBuilder.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+        for (Map.Entry<String, String[]> entry : parametersMap.entrySet()) {
+            String[] values = entry.getValue();
+            if (values != null && values.length > 0) {
+                specification = specification.and(specificationProviderManager
+                        .getSpecificationProvider(entry.getKey())
+                        .getSpecification(values));
+            }
+        }
+
```
Comment (beatarjn) [inline src/main/java/com/marek/onlinebookstore/repository/book/BookSpecificationBuilder.java]: Remove empty line

**Review by beatarjn — CHANGES_REQUESTED** [review]

**Review by MateuszRostek — APPROVED** [review]

**Pull request merged**